### PR TITLE
Add modern-webapp detection in git-service

### PR DESCRIPTION
**Pull request by rohitkrai03**
Related Bug - https://jira.coreos.com/browse/ODC-2065

This PR -
- Adds new logic to differentiate between modern web app and nodejs app.
- Now if `package.json` is present in a repo, build tool fetches the contents of the `package.json` and checks for dependencies if it is a modern web app like `react`, `angular`, `vue`.
- Based on the above logic it sorts the build tools where `modern-webapp` is added before `nodejs` in the list.
- Added unit tests for all the new scenarios.

Gif - 
ezgif com-video-to-gif (2)

### CODE REVIEW

File: frontend/packages/git-service/src/services/base-service.ts [hunk 1]
```diff
@@ -37,12 +37,19 @@ export abstract class BaseService {
...
   async detectBuildType(): Promise<BuildType[]> {
     try {
       const fileList = await this.getRepoFileList();
-      return detectBuildType(fileList.files);
+      let isWebApp = false;
+      if (fileList.files.includes('package.json')) {
+        const packageJsonContent = await this.getPackageJsonContent();
+        isWebApp = isModernWebApp(packageJsonContent);
+      }
+      return detectBuildType(fileList.files, isWebApp);
```
Comment (andrewballantyne) [inline frontend/packages/git-service/src/services/base-service.ts]: How about:
```
const buildTypes = detectBuildType(fileList.files);
if (fileList.files.includes('package.json')) {
  const packageJsonContent = await this.getPackageJsonContent();
  if (isModernWebApp(packageJsonContent)) {
    return buildTypes.sort((a, b) => {
      if (b.buildType === 'modern-webapp') return 1;
      if (a.buildType === 'modern-webapp') return -1;
      return 0;
    });
  }
}
return buildTypes;
```
That way we keep `detectBuildType` pure and leave ourselves open to move this logic elsewhere if we need to address other lib discrepancies.

Obviously you'd need to clean up `detectBuildType`.

**Comment by andrewballantyne** [comment]
Not that is overly matters - I just feel methods that return an array shouldn't have a singular name. `detectBuildType` should either return singular or be called `detectBuildTypes` imo.

File: frontend/packages/git-service/src/services/base-service.ts [hunk 2]
```diff
@@ -37,12 +37,19 @@ export abstract class BaseService {
   // Checks if dockerfile exist in the repo and returns dockerfile content
   abstract async getDockerfileContent(): Promise<string>;
 
+  abstract async getPackageJsonContent(): Promise<string>;
```
[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/git-service/src/services/base-service.ts]: Updated all the comments into JSDoc and added one for the new method.

File: frontend/packages/git-service/src/services/base-service.ts (hunk 1 above)
Comment (rohitkrai03) [reply frontend/packages/git-service/src/services/base-service.ts]: Yeah, that makes sense. Moved the logic of sorting based on modern web app to base service method.

**Comment by rohitkrai03** [comment]
> Not that is overly matters - I just feel methods that return an array shouldn't have a singular name. `detectBuildType` should either return singular or be called `detectBuildTypes` imo.

Updated the name to `detectBuildTypes`

**Pull request merged**